atomdb/base.py

```diff
@@ -318,6 +424,194 @@ def __get__(self, obj: T, cls: Optional[Type[T]] = None):
         raise NotImplementedError
 
 
+def generate_getstate(cls: Type["Model"], include_defaults: bool = True) -> GetStateFn:
+    """Generate an optimized __getstate__ function for the given model.
+
+    Parameters
+    ----------
+    cls: Type[Model]
+        The clase to generate a getstate function for.
+
+    Returns
+    -------
+    result: GetStateFn
+        A function optimized to generate the state for the given model class.
+
+    """
+    template = [
+        "def __getstate__(self, scope=None):",
+        "scope = scope or {}",
+        "scope[self.__ref__] = self",
+        "state = {",
+    ]
+    default_flatten = cls.serializer.flatten
+    members = cls.members()
+    namespace = {
+        "default_flatten": default_flatten,
+    }
+    for f in cls.__fields__:
+        # Since f is potentially an untrusted input, make sure it is a valid
+        # python identifier to prevent unintended code being generated.
+        if not f.isidentifier():
+            raise ValueError(f"Field '{f}' cannot be used for code generation")
+        m = members[f]
+        meta = m.metadata or {}
+        flatten = meta.get("flatten", default_flatten)
+        if flatten is default_flatten:
+            if is_primitive_member(m):
+                expr = f"self.{f}"
+            else:
+                expr = f"default_flatten(self.{f}, scope)"
+        else:
+            namespace[f"flatten_{f}"] = flatten
+            expr = f"flatten_{f}(self.{f}, scope)"
```

Comment:
I guess the cattr trick could be used here. If you are willing to go even further you could even consider sticking those into the function constant and using LOAD_CONST (but you will need bytecode for that). I assume it could be faster but I have not run any benchmark.

However you should document clearly that people won't be able to monkeypatch things the way they would expect to be able to do it usually but it can be a cool optimization.

Reply:
Will have to try this.. do you know of any examples that do this?  

To allow monkey patching in testing, I'm thinking of adding some sort of method to regenerate the functions (which would then update any const references).

Reply:
I don't think bytecode has any example of this kind but I know some users do mess with constants. I will try to spin up a small example for you.

Reply:
Here you go, I will probably add it to the bytecode repo too.

````python
"""Show how to turn a global attr into a constant.

This can allow to embed non-constant object in constant for optimization
purposes.

"""
from bytecode import Bytecode, Instr

def a():
    return 1


def b():
    return a()

def c():
    return a()


code = Bytecode.from_code(b.__code__)
for i, inst in enumerate(code):
    if inst.name == "LOAD_GLOBAL" and inst.arg == "a":
        code[i] = Instr("LOAD_CONST", a)

b.__code__ = code.to_code()

del a

print(b() == 1)

c()
```

Reply:
Thanks for the example.  Added it, hopefully doing that right.  It seems to not make too much of a difference in the benchmarks but I suppose that's because most of the time it can eliminate loading altogether.  Will try and see what it does in a real app.

Reply:
It may also be that the overhead of loading is tiny compared to the cost of flattening and you cannot really see it. But if cattr sees a win you should also be able to.